### DISCUSSION TITLE
General thoughts interface ideas/concerns/discussion

### DISCUSSION THREAD

**Discussion by WillAdams** [IndiePython/nodezator#6] [General]
There's already been a bit of discussion at: https://old.reddit.com/r/Python/comments/vf6h1z/nodezator_new_python_node_editor_released_to/

Here are some things to begin:

 - could right-click also be used as a contextual menu?
 - make it easier to break node connections? Perhaps right-clicking on a connecting line could afford the option of breaking it?
 - would double-clicking make sense as an interface option? It would feel a bit more natural for editing text rather than having to use the keyboard command ctrl t --- also might make a good interface option for breaking a connection
 - file hierarchy and naming conventions for a use adding nodes and sharing them?

**Comment by KennedyRichard** [comment]
> file hierarchy and naming conventions for a use adding nodes and sharing them?

This is already exists and is a simple as it gets: **you just need to share the folder with your nodes**.

For instance, I made the example nodes for working with Pillow [available here](https://github.com/KennedyRichard/pillow_example_node_pack).

Here are the steps to use the nodes:

- since the node pack relies on pillow, you need to install pillow in your Python;
- just download the folder anywhere you want in you system; I strongly recommend having a folder in your user folder called "node_packs" and save the node you downloaded there;
- then create a new nodezator file in which you want to use the nodes (or just open an existing one);
- then just click the command **Graph > Select node packs** on the menubar and select that folder and all the nodes will be available;

In other words, it is pretty straightforward: just download the nodes and enable them inside the app.

**Comment by KennedyRichard** [comment]
> and sharing them?

Sharing the nodes will be exactly like I just described: just upload them to your github and share the link with people on the internet, but there will be an additional tool:

As we briefly discussed on discord, I'm thinking of creating a website and put it on https://gallery.nodezator.com that works like a google for nodes and a node gallery, where people will be able to:
- submit description, screenshots and links to download their nodes;
- search for node packs...
    - by text
    - by author name: for instance, nodes by Mark, nodes by Karol, etc.
    - by tags/categories: for instance, "image editing", "3D modelling", etc.
- them to download them, they just click the link to the github repo where the node pack is available, download and install the nodes like in the previous comment.

**Comment by KennedyRichard** [comment]
I think I'll add all those steps/instructions in the manual, including inside the app, so people know how to use it. Do you think all the steps are ok? Are they simple enough?

I think having people share their nodes as repos on github is the best solution, because it is something which is straightforward and decentralized. If I were to create a centralized website or database, people would be dependent on the online service be always the case.

I've seen projects shutdown before and leave people unable to access essential resources. CTAN for instance had to shutdown new user registrations, see what they say on their website:

> Due to a spammer attack we had to disable the registration of new users. We are working on a solution. Please be patient. Sorry for any inconvenience.

For Nodezator, I'd like node sharing/distribution to rely on github, this way all nodes will always be available no matter what happens to the project (though don't worry, the project is strong and came to stay, I'm just doing what I think is best). And we'll still have https://gallery.nodezator.com to share useful nodes node packs and help people find them, pointing people to the github repos where they can be downloaded.

**Comment by KennedyRichard** [comment]
> * could right-click also be used as a contextual menu?

Yes, I already intend to do so. The source code for some of the nodes, for instance, even have a right-click event listener, I just didn't implemented the context menu yet. The only thing preventing me from implementing them is to know what to put on them.

Some objects even have context menu's already. For instance, input sockets created for subparameters (sockets created under a variable parameter like *args or **kwargs) already have a context menu, you can see it in actions in this [part of the video](https://youtu.be/GlQJvuU7Z_8?t=1108), because I already had some idea of which commands to put there.

In other words, context menus will sure be there. I'll create 02 new discussions later today:

- New discussion number 1: One where I list the changes planned for the next version.
- New discussion number 2: And other with a general list of things to be implemented, regardless of whether they will be implemented in the next version or the one after, etc.;

This way we'll be able to keep track of things to be done and when they will be arriving, etc.

> * make it easier to break node connections? Perhaps right-clicking on a connecting line could afford the option of breaking it?

This sounds good, I too think right-clicking should display the option to do so in a context menu. I'll add it to the to-do list for the next version which I mentioned above as the **New discussion number 1**.

> * would double-clicking make sense as an interface option? It would feel a bit more natural for editing text rather than having to use the keyboard command ctrl t --- also might make a good interface option for breaking a connection

I'm thinking of adding the double-clicking to edit text blocks to the **New discussion number 1** later today as well. I'm not sure it would be good interface for breaking connections though. It could be enabled though, but let users opt-out of it in the user preferences, I'm not sure. Let me ponder about it a bit more.

**Comment by zeckstein** [comment]
Resizable main window a consideration?

**Reply by WillAdams** [reply]
Yes, it would be a huge thing, and would be a prime impetus in my shifting working to my desktop w/ 4K display (usually I use a small tablet).

**Reply by KennedyRichard** [reply]
Much more than a consideration, actually. Resizable window is both desirable and useful and is listed among the things to be added to Nodezator. I have prioritized bug fixing and issues since the release this past weekend, but I'll soon start a discussion listing the planned features/roadmap for the next versions.

**Reply by KennedyRichard** [reply]
@WillAdams , the resizable window implementation is going smoothly, but I also wanted to tell you that though you'll certainly benefit from the feature, since you'll be able to increase the amount of the screen used, I'm not sure the app will be performant enough if you use the full 4k dimensions. I commented on it here https://github.com/KennedyRichard/nodezator/discussions/14#discussioncomment-3057189 but also wanted to talk to you here, since you expressed your intentions of using the app in your 4k display.

Again, you'll still benefit and use a bigger window, I'm just not sure you'll be able to use the **entire** screen. Well, we'll see when it is implemented and you test it. I just thought it is would be fair to warn you. Please, keep this in mind.

**Comment by WillAdams** [comment]
Some other thoughts:

 - it would be nice to have a more pleasing appearance for the lines --- curves are a typical representation --- could we get at least an option for that?
 - What is the underlying appearance/representation for on-screen? I see that there's an "Export as Image" command under File --- Would it be possible to have an export option for a vector representation? SVG would be great!
 - zoom in/out?
 - possible to improve the custom UI elements so that they respond to typical inputs such as double-clicking (to select text) or tapping Enter/Return to accept and close the dialog box? Map Esc to Cancel, &c. (Is this something you made for this project? If so, should it be split off so that other folks can use it in appropriate projects and maybe it'll get some traction for improvements?)
 - would it be workable to embed this in an application as a scripting/extension language? FreeCAD comes to mind.

**Reply by KennedyRichard** [reply]
> What is the underlying appearance/representation for on-screen? I see that there's an "Export as Image" command under File --- Would it be possible to have an export option for a vector representation? SVG would be great!

On screen the graphics are represented by pygame.Surface objects, a pygame class.

The graph can be exported to .html, .svg or .png. All you need to do is change the extension of the filename. [Here's a demonstration](https://youtu.be/GlQJvuU7Z_8?t=1211), from the 20:11 mark in the video.

**Reply by WillAdams** [reply]
Outstanding! (Obviously I need to re-watch the video and take some notes --- is there a transcript?)

Thanks! (and my thanks for your patience and forbearance)

**Reply by KennedyRichard** [reply]
> it would be nice to have a more pleasing appearance for the lines --- curves are a typical representation --- could we get at least an option for that?

> zoom in/out?

At the moment this isn't possible. The default pygame graphics backend doesn't offer an option to draw curves or zoom in/out. However, sometime in the future I should add the option to use pygame using OpenGL as the graphics backend. When this happens, both features will be way easier to implement.

Moreover, using curves for connections does look beautiful, but considering what I observed with Blender3D, some people disable the curves when their graphs get too big so the graph don't get too cluttered and some renowned authors like Barted Skorupa even recommend people to reduce their "noodle curving" by default. It seems  recent versions of Blender3D even have noodle curving default to 0 (thus using straight lines). I'm saying this just as an observation though, since I see no problem in implementing an equivalent "noodle curving" option in Nodezator once the OpenGL backend is implemented. It will take sometime until Nodezator gets to this point though, but I'll list your demand among things to implement, I think it is reasonable.

**Reply by KennedyRichard** [reply]
> Outstanding! (Obviously I need to re-watch the video and take some notes --- is there a transcript?)

Unfortunately no official transcript at the moment. I do have a lot of the video as text, since I usually write a draft for my videos, but it is not 100% precise because sometimes I make my own changes on the fly when recording voice/screen. But I'll fix the inaccuracies and make it available sometime. For my next videos I'll make sure every changes goes back to the draft, so that it can be published along with the video as a transcript.

> Thanks! (and my thanks for your patience and forbearance)

No problem, it is my honest pleasure to be able to help people. This is my stance no matter whether the person I'm talking with is a patron or not of the project. The fact that you are a patron though does result in a bit more of priority to your demand, though, not because paying makes you more important, but because you chose to contribute when you could've just used the software for free anyway. And your selfless noble contribution helps me by allowing me to allocate more time and effort to the project, thus helping everyone in the end.

Open source, when done well, is a positive feedback loop that benefits everyone. I hope you can keep contributing for a long time, but if for some reason you cannot anymore or decide it is not worthy, it will be okay as well, you'll be treated the same. But, as far as it depends on me I'll go the extra mile to show people that contributing to the Indie Python project is worthy.

> possible to improve the custom UI elements so that they respond to typical inputs such as double-clicking (to select text) or tapping Enter/Return to accept and close the dialog box? Map Esc to Cancel, &c. (Is this something you made for this project? If so, should it be split off so that other folks can use it in appropriate projects and maybe it'll get some traction for improvements?)

The tapping Enter/Return to accept and close the dialog box is possible and it is my desire to implement such things as well. Selecting text will take more time to implement, but is something I think is very important as well. I'll schedule those changes, but the text selection will take a bit of time.

What do you mean by "split off" here? Could you please elaborate?

> would it be workable to embed this in an application as a scripting/extension language? FreeCAD comes to mind.

Yes, it is possible. Not only FreeCAD, but any software that can use Python can use Nodezator as a node layout. Let me explain.

I think one of the best things about Nodezator is the fact that the node layouts can be exported as Python code, because, in the end, you can think of node layouts as just a bunch of Python functions mixed together. In that sense, any app that has access to Python one way or another can take advantage of Nodezator as a node editor, because it can simply make use of the exported script. Thus, everything that uses Python can benefit from Nodezator by using the exported script as a bridge.

In the [video I presented Nodezator](https://youtu.be/GlQJvuU7Z_8?t=1902), at the 31:42 mark, I even discuss this and give a demonstration where I use Nodezator to create a node layout representing an image filter and the layout (exported as a Python script) in a Tkinter app, to filter images.

I'm aware that Python can be used in FreeCAD, so using node layouts from Nodezator in FreeCAD should not only be doable, but relatively easy. Having all of this in mind, I plan to produce more demonstrations like that so that people are aware of Nodezator's advantages/capabilities. FreeCAD does look like a good candidate for being among those demonstrations.

If you have something specific in FreeCAD that you wish could be done with a node layout do share and we can see if we can come up with a mock up/mvp version of such scripting/extension as an starting point.

**Comment by WillAdams** [comment]
Another feature idea --- would it be desirable to add an option for a graphical icon for a node? The GraphSCAD folks do this of course, also the Nodebox folks, so it's probably more useful for the 3D/graphical work I'm trying to get done, but it might be useful for other things --- though it raises the question, "What does an algorithm look like?"

**Reply by KennedyRichard** [reply]
This is interesting. I thought of this as well, but since I didn't have a clear picture of this feature in mind I refrained from trying implementing anything of the sort.

You have already pointed out the apps you use on your workflow like GraphSCAD as a source of inspiration/example for feature you'd like on Nodezator, but I didn't have time yet to take a look into them. I'm sorry for this.

I'll take a look asap. I'm certain a quick google search will provide me all the info and resources I need, but if there's any resource in specific (video/article/link) you think I should also look into, please share.

After I'm a bit more acquainted with that app I'll share my thoughts with you, including your feature idea of add custom graphical representations for nodes.

My only rule of thum is that every feature that makes into Nodezator must be translatable into Python code, that is, it must be compatible with the **exporting as Python** feature, but custom graphical representations doesn't seem to cause any harm to this. I'll soon share my thoughts on it, say, in a couple of days.

**Reply by WillAdams** [reply]
No problem/hurry --- I'm grateful for what you've done (and need to find time to get some coding done myself, busy testing a new beta for my side gig).

Will definitely keep the exporting as Python aspect in mind for future feature suggestions.
